src/storage/v2/vertex_accessor.cpp

```diff
@@ -616,16 +619,40 @@ Result<EdgesVertexAccessorResult> VertexAccessor::InEdges(View view, const std::
   {
     auto guard = std::shared_lock{vertex_->lock};
     deleted = vertex_->deleted;
-    expanded_count = static_cast<int64_t>(vertex_->in_edges.size());
-    // TODO: a better filter copy
-    if (edge_types.empty() && !destination) {
-      in_edges = vertex_->in_edges;
+    if (hops_limit && hops_limit->IsUsed()) {
+      if (edge_types.empty() && !destination) {
+        // we need this check because if limit is reached but it is not passed we can still reach vertices without any
+        // neighbors
+        if (hops_limit->LeftHops() == 0 && static_cast<int64_t>(vertex_->in_edges.size()) > 0) {
+          hops_limit->limit_passed = true;
+        } else {
+          expanded_count = (std::min(hops_limit->LeftHops(), static_cast<int64_t>(vertex_->in_edges.size())));
+          hops_limit->IncrementHopsCount();
+          std::copy_n(vertex_->in_edges.begin(), expanded_count, std::back_inserter(in_edges));
+        }
+      } else {
+        for (const auto &[edge_type, to_vertex, edge] : vertex_->in_edges) {
+          hops_limit->IncrementHopsCount();
+          if (hops_limit->IsLimitReached()) break;
+          expanded_count++;
+          if (destination && to_vertex != destination_vertex) continue;
```

Comment:
This code is duplicated with below code. In next PR please consider to simplify this function. Lot of nested if conditions makes it complex to understand what's going on in the function.